server/src/main/scala/drt/server/feeds/test/TestAPIManifestFeedGraphStage.scala

```diff
@@ -0,0 +1,93 @@
+import akka.stream.{ActorMaterializer, Attributes, Outlet, SourceShape}
+import akka.util.Timeout
+import drt.shared.CrunchApi.MillisSinceEpoch
+import org.slf4j.{Logger, LoggerFactory}
+import passengersplits.parsing.VoyageManifestParser.VoyageManifests
+import services.SDate
+import services.graphstages.DqManifests
+import test.TestActors.ResetActor
+
+import scala.concurrent.Await
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+import scala.language.postfixOps
+
+
+class TestAPIManifestFeedGraphStage(actorSystem: ActorSystem) extends GraphStage[SourceShape[DqManifests]] {
+  implicit val system = actorSystem
+  implicit val materializer: ActorMaterializer = ActorMaterializer()
+  val log: Logger = LoggerFactory.getLogger(getClass)
+  val minCheckIntervalMillis = 3000
+
+  val out: Outlet[DqManifests] = Outlet("manifestsOut")
+  override val shape: SourceShape[DqManifests] = SourceShape(out)
+
+  var dqManifestsState: DqManifests = DqManifests("today", Set())
+  var lastFetchedMillis: MillisSinceEpoch = 0
+
+  val askableTestManifestsActorActor: AskableActorRef = actorSystem.actorOf(Props(classOf[TestManifestsActor]), s"TestActor-APIManifests")
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = {
+    new GraphStageLogic(shape) {
+      setHandler(out, new OutHandler {
+        override def onPull(): Unit = {
+          fetchAndUpdateState()
+          pushManifests()
+        }
+      })
+
+      def pushManifests(): Unit = {
+        if (isAvailable(out)) {
+
+          push(out, dqManifestsState)
+
+          dqManifestsState = dqManifestsState.copy(manifests = Set())
+        }
+      }
+    }
+  }
+
+  def fetchAndUpdateState() = {
+    val nowMillis = SDate.now().millisSinceEpoch
+    val millisElapsed = nowMillis - lastFetchedMillis
+    if (millisElapsed < minCheckIntervalMillis) {
+      val millisToSleep = minCheckIntervalMillis - millisElapsed
+      log.info(s"Minimum check interval ${minCheckIntervalMillis}ms not yet reached. Sleeping for ${millisToSleep}ms")
+      Thread.sleep(millisToSleep)
+    }
+
+    implicit val timeout: Timeout = Timeout(300 milliseconds)
+    dqManifestsState = Await.result(askableTestManifestsActorActor.ask(GetManifests).map {
+      case VoyageManifests(manifests) =>
+        log.info(s"Got these manifests from the actor: $manifests")
+        dqManifestsState.update("today", manifests)
+    }, 10 seconds)
+    lastFetchedMillis = nowMillis
+  }
+}
+
+case object GetManifests
+
+class TestManifestsActor extends Actor with ActorLogging {
+
+  var testManifests: VoyageManifests = VoyageManifests(Set())
+
+  override def receive = {
+    case vms: VoyageManifests =>
+      log.info(s"Got these VMS: $vms")
+      testManifests = testManifests.copy(manifests = testManifests.manifests ++ vms.manifests)
+
+    case GetManifests =>
+
+      sender ! VoyageManifests(testManifests.manifests)
```

Comment:
do we want this test actor to clear the test manifests as soon as it has sent them? or are there cases where we'd want it to send the same ones again multiple times?

Reply:
This replicates the behaviour of the real one - it will keep sending everything it has right?